typing/typedecl.ml

```diff
@@ -226,6 +253,34 @@ let make_constructor env type_path type_params sargs sret_type =
       widen z;
       targs, Some tret_type, args, Some ret_type
 
+(* Check that the argument to a GADT constructor is compatible with unboxing
+   the type, given the existential variables introduced by this constructor. *)
+let rec check_unboxed_gadt_arg loc ex env ty =
+  match get_unboxed_type_representation env ty with
+  | Some {desc = Tvar _ as d; _} ->
+    let f t = (Btype.repr t).desc = d in
+    if List.exists f ex then
+      raise(Error(loc, Wrong_unboxed_type
+                    "it might contain both float and non-float values"))
+  | Some {desc = Tarrow _ | Ttuple _ | Tpackage _ | Tobject _ | Tnil
+                 | Tvariant _; _} ->
+    ()
+    (* A comment in [Translcore.transl_exp0] claims the above cannot be
+       represented by floats. *)
+  | Some {desc = Tconstr (p, args, _); _} ->
+    let tydecl = Env.find_type p env in
+    assert (not (Builtin_attributes.has_unboxed tydecl.type_attributes));
```

Comment:
I don't see immediately why the argument could not be an unboxed type (if it is, the check should fail, but not with an assertion failure).